components/script/dom/document.rs

```diff
@@ -418,6 +375,10 @@ pub struct Document {
     delayed_tasks: DomRefCell<Vec<Box<dyn TaskBox>>>,
     /// https://html.spec.whatwg.org/multipage/#completely-loaded
     completely_loaded: Cell<bool>,
+    /// List of shadow roots bound to the document tree.
+    shadow_roots: DomRefCell<Vec<Dom<ShadowRoot>>>,
```

Comment:
Given this doesn't need any order, it may as well be a HashSet, but maybe followup if it's hard.